cluster/cluster.go

```diff
@@ -0,0 +1,431 @@
+		lock      sync.Mutex
+		instances = c.env.Instances()
+		multiErr  xerrors.MultiError
+	)
+
+	// setup the instances, in parallel
+	for idx := range instances {
+		inst := instances[idx]
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			if err := inst.Setup(svcBuild, svcConf); err != nil {
+				lock.Lock()
+				multiErr = multiErr.Add(err)
+				lock.Unlock()
+				return
+			}
+
+			lock.Lock()
+			c.spares = append(c.spares, inst)
+			lock.Unlock()
+		}()
+	}
+	wg.Wait()
+
+	return c.markStatus(ClusterStatusSetup, multiErr.FinalError())
+}
+
+func (c *m3dbCluster) Initialize(numNodes int) ([]env.M3DBInstance, error) {
+	if c.Status() != ClusterStatusSetup {
+		return nil, errClusterUnableToInitialize
+	}
+
+	numSpares := len(c.spares)
+	if numSpares < numNodes {
+		return nil, errInsufficientCapacity
+	}
+
+	usedInstances := c.spares[:numNodes]
+	instances := make([]services.PlacementInstance, 0, numNodes)
+	for _, inst := range usedInstances {
+		c.usedInstances[inst.ID()] = inst
+		instances = append(instances, inst)
+	}
+	c.spares = c.spares[numNodes:]
+
+	psvc := c.placementSvc
+	_, err := psvc.BuildInitialPlacement(instances, c.copts.NumShards(), c.copts.Replication())
+	if err != nil {
+		return nil, err
+	}
+
+	c.statusLock.Lock()
+	c.status = ClusterStatusInitialized
+	c.statusLock.Unlock()
+	return usedInstances, nil
+}
+
+func (c *m3dbCluster) AddInstance() (env.M3DBInstance, error) {
+	if status := c.Status(); status != ClusterStatusRunning && status != ClusterStatusInitialized {
+		return nil, errClusterUnableToAlterPlacement
+	}
+
+	numSpares := len(c.spares)
+	if numSpares < 1 {
+		return nil, errInsufficientCapacity
+	}
+	inst := c.spares[0]
+	psvc := c.placementSvc
+	_, err := psvc.AddInstance([]services.PlacementInstance{inst})
+	if err != nil {
+		return nil, err
+	}
+	c.usedInstances[inst.ID()] = inst
+	c.spares = c.spares[1:]
+	return inst, nil
+}
+
+func (c *m3dbCluster) RemoveInstance(i env.M3DBInstance) error {
+	if status := c.Status(); status != ClusterStatusRunning && status != ClusterStatusInitialized {
+		return errClusterUnableToAlterPlacement
+	}
+
+	if _, ok := c.usedInstances[i.ID()]; !ok {
+		return errInstanceNotInUse
+	}
+	psvc := c.placementSvc
+	_, err := psvc.RemoveInstance(i.ID())
+	if err != nil {
+		return err
+	}
+	delete(c.usedInstances, i.ID())
+	c.spares = append(c.spares, i)
+	return nil
+}
+
+func (c *m3dbCluster) ReplaceInstance(oldInstance env.M3DBInstance) (env.M3DBInstance, error) {
+	if status := c.Status(); status != ClusterStatusRunning && status != ClusterStatusInitialized {
+		return nil, errClusterUnableToAlterPlacement
+	}
+
+	if _, ok := c.usedInstances[oldInstance.ID()]; !ok {
+		return nil, errInstanceNotInUse
+	}
+
+	numSpares := len(c.spares)
+	if numSpares < 1 {
+		return nil, errInsufficientCapacity
+	}
+
+	spareInstance := c.spares[0]
+	psvc := c.placementSvc
+	_, err := psvc.ReplaceInstance(oldInstance.ID(), []services.PlacementInstance{spareInstance})
```

Comment:
this could stuck in a case where c.spares[0] could not be used to replace the oldInstance, depending on how you config the placement options, if you require the replace to be a complete replace it's possible that c.spares[0] can't take all the shards from oldInstance due to rack conflicts.

The placement service also supports weight matching if configured, for example if oldInstance's weight is 100, all the spare instances are of weight 20, 5 spares instances could be picked for the replacement.